Fix metadata format

SantiagoTorres opened the pull request:
The current state of the in-toto classes do not match with the description in the specification (closes #98) 

Please review commit by commit, so as to make historical sense **and** to simplify your review process.

You will probably notice some OOP abuse, and other horrendous things (I apologize to the language puritans out there in advance). These are the things that I don't pledge allegiance to, but I included anyway (I was hoping to ticketize/discuss them):

- The abuse of OOP (also known as, "**why is there a mixin in my diff**")
- Is that a custom __getattr__ implementation?
- The wonky adapter pattern in the constructor
- The if-else clauses in the constructor (specially for links) so as to be too smart around the code throwing various types at it.

Most of these were added to introduce a "working copy" in under a 300 line diff. I hope we can slowly pour in upgrades and cleanups to get rid of the mixin, the "custom builtins" and the "why does this code try handle 3 different things at the same time" instances.

SantiagoTorres committed:
ENH: .gitignore: ignores .travis-solo

SantiagoTorres committed:
ENH: in-toto/models: update structure to match spec

The models were restructured to match the specification. Specifically,
the fields were separated into two groups:

- Signed: the fields from which the signature was computed (or is to be
computed)
- Signatures: The signatures over the signed fields.

To do this, I introduced a custom __getattr__ module to reduce the size
of the patch and make most of the tests pass. This is not the behavior
that we want in the long run, but it will let us slowly migrate to a
solid solution.

SantiagoTorres committed:
ENH: in-toto/in-toto-sign: support format

Since the metadata format change, a couple of guard clauses on the
deserialized json would fail (given that they assume the old schema).
Fix these guards to be aware of this.

SantiagoTorres committed:
ENH: in-toto/runlib: support new format

runlib instantiates link metadata in many of its methods. Due to the
nature of the initializer, the fields of the metadata are not set
properly: they are not set in the nested 'signed' property, but
"besides" it. This could've been fixed with a custom setattr for the
Metablock class, but that solution was its own can of worms.

SantiagoTorres committed:
ENH: test/demo-files: update format

updated the format of the demo files, which consequentially requires
re-signing.

SantiagoTorres committed:
ENH: test: update test setup and verifications

Due to the new format specification, some of the test SetUp calls do not
initialize metadata properly. Update these calls to properly initialize
the test metadata -- this includes inline metadata.

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+  def __getattr__(self, item):
+    # if the item was malformed, then return none
+    if item == 'signatures' or item == 'signed':
+        return None
```
Fix 4 space indentation

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+    signature = securesystemslib.keys.create_signature(key, repr(self.signed))
+    self.signatures.append(signature)
+
+  """ FIXME: This is mostly sintactic sugar and backwards compatibility stuff
```
typo: sintactic

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+      self.signatures = kwargs['signatures']
+      del kwargs['signatures']
+
+    signable = self.get_signable()
```
Should we define `get_signable` in the base class `Metablock` as well? Something like:

```python
def get_signable(self):
  raise NotImplementedError()
```

SantiagoTorres replied:
Yup, sounds  like a good idea.

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
   file """
   def __repr__(self):
-    return canonicaljson.encode_pretty_printed_json(attr.asdict(self))
+    return '{{"signed": {}, "signatures": {}}}'.format(self.signed,
```
Mind adding a comment about double braces and replacement fields and?

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+
+  """ FIXME: This is mostly sintactic sugar and backwards compatibility stuff
+  """
+  def __getattr__(self, item):
```
Please provide a docstring with more details

lukpueh replied:
I wonder if we shouldn't make the effort and fix all the unit tests, etc. now at a time where we don't really care for backwards compatibility?
Also, overriding `__getattr__` is just part of the story, as can be seen in  https://github.com/in-toto/in-toto/pull/120/commits/c2f703085eeadd4a7d8dcc5f26a8307ed404f977

SantiagoTorres replied:
Yeah, I had a version in which I also overwrote `__setattr__` but it felt particularly insulting.

I can do a bulk replace of these changes all throughout the tests, but I probably want to separate the bulk rename of the rest of the modules to a separate PR.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -275,7 +295,7 @@ def _validate_steps_and_inspections(self):
 
 @attr.s(repr=False, init=False)
-class Step(models__common.Metablock):
+class Step(models__common.Signable):
```
Step is not "signable". Please replace base class `Signable` with `ValidationMixin`.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -383,7 +403,7 @@ def _validate_expected_command(self):
 
 @attr.s(repr=False, init=False)
-class Inspection(models__common.Metablock):
+class Inspection(models__common.Signable):
```
Inspection is not "signable". Please replace base class `Signable` with `ValidationMixin`.

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+      else:
+        self.signed = signable(**kwargs['signed'])
+    else:
+        self.signed = signable()
```
Fix 4 space indentation

lukpueh commented on in_toto/models/common.py:
```diff
@@ -69,47 +136,16 @@ def validate(self):
 
 @attr.s(repr=False, init=False)
-class Signable(Metablock):
+class Signable(ValidationMixin):
```
Class docstring and header block need to be updated as well.

lukpueh commented on in_toto/models/common.py:
```diff
@@ -69,47 +136,16 @@ def validate(self):
-        raise SignatureVerificationError("Invalid signature")
 
+  """Objects with base class Signable have a __repr__ method
+  that returns a canonical pretty printed JSON string and can be dumped to a
```
I'd remove the "and can be dumped to a file".

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -77,31 +77,10 @@ class Layout(models__common.Signable):
-        relativedelta(months=1)).strftime("%Y-%m-%dT%H:%M:%SZ"))
-
-    self.validate()
+  def get_signable(self):
```
This needs a docstring

lukpueh commented on in_toto/models/common.py:
```diff
@@ -32,18 +32,85 @@
+  """
+
+  def __init__(self, **kwargs):
+
```
Please add a docstring that explains your adaptor wizardry.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -47,8 +47,8 @@
-@attr.s(repr=False, init=False)
-class Layout(models__common.Signable):
+
+class Layout(models__common.Metablock):
```
Doesn't below docstring belong to `_Layout` now? I'd rather explain here how and why `Layout` holds a `_Layout` in its `signed` property (inherited from the `Metablock` base class).

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -120,15 +99,21 @@ def read_from_file(filename='root.layout'):
+        data['signed'] = _Layout()
+
+    for step_data in data['signed'].get("steps"):
+      # FIXME: is this really needed?
```
I think so yes, since we don't want to assign a list of step dictionaries to the `steps` attribute but a list of `Step` objects.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -203,6 +193,36 @@ def import_step_metadata_from_files_as_dict(self):
 
+
+@attr.s(repr=False, init=False)
+class _Layout(models__common.Signable):
```
What about a more descriptive name, like e.g. `LayoutSignable`, or `LayoutPayload`? What do you think?

SantiagoTorres replied:
That's probably clearer in the long term. I adopted this approach because of the `_` prefix, but it's probably not that clear/useful. Will change.

lukpueh commented on in_toto/models/link.py:
```diff
@@ -26,8 +26,7 @@
 
-@attr.s(repr=False, init=False)
-class Link(models__common.Signable):
+class Link(models__common.Metablock):
```
See comment about `Layout` vs. `_Layout` docstrings. IMO same applies here.

lukpueh commented on in_toto/models/link.py:
```diff
@@ -60,25 +59,22 @@ class Link(models__common.Signable):
+    super(Link, self).__init__(**kwargs)
+
+  def get_signable(self):
+    """ this method is used by the base class's constructor to obtain the
```
We usually start docstrings with a capital letter.

lukpueh reviewed (changes requested):
Thanks for the PR, @SantiagoTorres! I think it's an elegant solution.

As a few general notes:
- I think the validator mixin is alright, but I would probably disassociate it from `Signable` and use it independently
- I think we should get rid of the custom `getattr` now, as it does not really help with backwards compatibility
- The manifestations of your *wonky adapter pattern* need some more comments, and if I understood your correctly in that it's just a temporary solution, I would also add some inline `TODO`s or so.

lukpueh commented:
Also note that the PR breaks the [demo](https://github.com/in-toto/in-toto/tree/develop/demo) (and the [opensuse-demo](https://github.com/in-toto/in-toto/tree/develop/demo-opensuse)).

Commit:
Adapt demos to match signed/signature structure

This commit updates the demo layout creation scripts to match
the new metadata structure, where the signable parts of Layouts
and Links are contained in a top level property called
"signed" and the "signatures" property is on the same level.

lukpueh commented:
Just spotted a couple of more indentation mismatches (4 instead of 2 spaces). Please fix.

Commit:
Add unittest and fix indent in in_toto_sign

Add a unittest to prevent coverage decline.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -120,15 +99,21 @@ def read_from_file(filename='root.layout'):
-    for step_data in data.get("steps"):
+
+    if 'signed' not in data:
+        data['signed'] = _Layout()
```
I think you could leave this out (which would also prevent coverage decline), as you do the same thing in `common.Metablock`'s constructor anyway, right? https://github.com/in-toto/in-toto/blob/c1e0cf122f2e1faf6583784517a23cb752a0fc77/in_toto/models/common.py#L56-L57


You could wrap the lines below in a `if "signed" in data:`, or something like `data.get("signed", {}).get("steps")` which I like but others might find ugly.

SantiagoTorres replied:
Sure will do, but I'm less worried about losing 0.01% of coverage than having an increasingly convoluted constructor.

 I do think that this is bringing up the bad coupling patterns that we have on our constructors though. This spaghetti code is just reflecting the fact that we'd rather use .read and inline the json rather than instantiating metadata objects all throughout the code.

lukpueh commented on in_toto/models/layout.py:
```diff
@@ -177,12 +162,17 @@ def import_step_metadata_from_files_as_dict(self):
+
+          if 'signed' not in link_obj:
+            raise in_toto.exceptions.LinkNotFoundError("Invalid format. '{0}'"
+                " is not a valid in-toto metadata file.".format(filename))
```
Can you add a unittest for this line?

SantiagoTorres replied:
interesting you proposed adding `data.get("signed", {}).get("steps")` above but not here. Any reasoning behind this?

I'm reluctant to add a UT on this line as this feels like code that should be refactored (why don't we have a schema run on this?) why do we use schema anyway?

lukpueh replied:
Only negligible reasoning really. :) But yes, I thought about proposing the same solution here as well. What do you think?

The reason why we don't schema validate her is because we don't have link schema validation yet (which we should add https://github.com/in-toto/in-toto/issues/122).

SantiagoTorres committed:
DOC: models:common: add docstrings to metablock

Some of the methods of the metablock were missing docstrings. I also
sneaked in a "notimplementederror" exception in case someone wants to
instantiate a metablock.

SantiagoTorres committed:
DOC: update classnames and docstrings for Layout

The Layout class now uses clearer class names. This patch adds the
docstrings to the methods that were missing it.

SantiagoTorres committed:
MAINT: models:layout fix step/inspec, inherittance

Inspections are steps were listed as inheritted from a signable class.
This is not accurate, as they aren't strictly speaking a "signable".
Make Link and Layout inherrit from the ValidationMixin instead.

SantiagoTorres committed:
FIXUP: link: rename _Link to LinkSignable

Change the name of the _Link signable to match the convention used in
the layout module. This also fixes docstrings.

SantiagoTorres committed:
MAINT: layout: removes unused layoutsignable call

A layout signable call used to be done in the read method. Given that
this is done by the constructor anyway, it doesn't need to be called on
the read.

SantiagoTorres committed:
MAINT: models:Signable: removes unused constructor

The signable constructor did nothing at all (but to add useless code).
Remove the constructor and propagate the changes throughout the code.

SantiagoTorres committed:
ENH: models: layout: simplify schema checks

Some schema checks in the layout constructor were redundant or overly
complicated. SImplify the schema checks in the read() and
import_step_metadata_from_files_as_dict() methods.

SantiagoTorres committed:
TEST: models:common: add test suite for Metablock

The metablock class has code that remains untested. Adding a metablock
test suite that ensure that Metablock remains abstract takes care of
this.

Commit:
Minor typo and whitespace fixes

lukpueh commented:
Thanks @SantiagoTorres!
